Source/WebKit/Platform/IPC/Decoder.cpp

```diff
@@ -27,58 +27,52 @@
 #include "Decoder.h"
 
 #include "ArgumentCoders.h"
-#include "DataReference.h"
 #include "Logging.h"
 #include "MessageFlags.h"
 #include <stdio.h>
 #include <wtf/StdLibExtras.h>
 
 namespace IPC {
 
-static const uint8_t* copyBuffer(const uint8_t* buffer, size_t bufferSize)
+static const uint8_t* copyBuffer(DataReference buffer)
 {
     uint8_t* bufferCopy;
+    const size_t bufferSize = buffer.size_bytes();
     if (!tryFastMalloc(bufferSize).getValue(bufferCopy)) {
         RELEASE_LOG_FAULT(IPC, "Decoder::copyBuffer: tryFastMalloc(%lu) failed", bufferSize);
         return nullptr;
     }
 
-    memcpy(bufferCopy, buffer, bufferSize);
+    memcpy(bufferCopy, buffer.data(), bufferSize);
     return bufferCopy;
 }
 
-std::unique_ptr<Decoder> Decoder::create(const uint8_t* buffer, size_t bufferSize, Vector<Attachment>&& attachments)
+std::unique_ptr<Decoder> Decoder::create(DataReference buffer, Vector<Attachment>&& attachments)
 {
-    ASSERT(buffer);
-    if (UNLIKELY(!buffer)) {
-        RELEASE_LOG_FAULT(IPC, "Decoder::create() called with a null buffer (bufferSize: %lu)", bufferSize);
-        return nullptr;
-    }
-    return Decoder::create(copyBuffer(buffer, bufferSize), bufferSize, [](const uint8_t* ptr, size_t) { fastFree(const_cast<uint8_t*>(ptr)); }, WTFMove(attachments)); // NOLINT
+    return Decoder::create({ copyBuffer(buffer), buffer.size() }, [](DataReference buffer) { fastFree(const_cast<uint8_t*>(buffer.data())); }, WTFMove(attachments)); // NOLINT
```

Comment:
We shouldn't need a const_cast here.

Reply:
The compiler says we do need it:

```
OpenSource/Source/WebKit/Platform/IPC/Decoder.cpp:52:94: error: no matching function for call to 'fastFree'
    return Decoder::create({ copyBuffer(buffer), buffer.size() }, [](DataReference buffer) { fastFree(buffer.data()); }, WTFMove(attachments)); // NOLINT
                                                                                             ^~~~~~~~
In file included from WebKitBuild/Debug/DerivedSources/WebKit/unified-sources/PlatformUnifiedSource3.cpp:1:
In file included from OpenSource/Source/WebKit/Platform/IPC/DaemonEncoder.cpp:26:
In file included from OpenSource/Source/WebKit/config.h:42:
WebKitBuild/Debug/usr/local/include/wtf/FastMalloc.h:103:25: note: candidate function not viable: no known conversion from 'std::span<const unsigned char, 18446744073709551615>::pointer' (aka 'const unsigned char *') to 'void *' for 1st argument
WTF_EXPORT_PRIVATE void fastFree(void*);
                        ^
1 error generated.
```

Or are you thinking we should change the definition of `DataReference` to `std::span<uint8_t>`?

Reply:
Chris and I chatted in person, and changing `DataReference` to `std::span<uint8_t>` is a bigger change than affects more code than just this patch, and may have other implications, so I think it would be best (if we decide to do it) to land as a separate patch.